### PR TITLE
AMDGPU: Handle legal v2f16/v2bf16 atomicrmw fadd for global/flat

### PR DESCRIPTION
**Pull request by arsenm**
Unlike the existing fadd cases, choose to ignore the requirement for
amdgpu-unsafe-fp-atomics in case of fine-grained memory access. This
is to minimize migration pain to the new atomic control metadata. This
should not break any users, as the atomic intrinsics are still
directly consumed, and clang does not yet produce vector FP atomicrmw.

### CODE REVIEW

**Comment by arsenm** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95394?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#95396** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95396?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#95395** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95395?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#95394** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95394?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#95393** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95393?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Target/AMDGPU/FLATInstructions.td [hunk 1]
```diff
@@ -1669,13 +1670,16 @@ defm : FlatSignedAtomicPatWithAddrSpace <"FLAT_ATOMIC_ADD_F32", "int_amdgcn_flat
 }
 
 let OtherPredicates = [HasAtomicFlatPkAdd16Insts] in {
+// FIXME: These do not have signed offsets
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/FLATInstructions.td:1673]: Can you just use FlatAtomicPat?

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 2]
```diff
@@ -15931,6 +15931,26 @@ static OptimizationRemark emitAtomicRMWLegalRemark(const AtomicRMWInst *RMW) {
          << " operation at memory scope " << MemScope;
 }
 
+static bool isHalf2OrBFloat2(Type *Ty) {
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:15934]: Does the underlying type really matter? Is 2 x 16-bit type sufficient?

File: llvm/lib/Target/AMDGPU/FLATInstructions.td (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/FLATInstructions.td:1673]: Yes, but I was planning on copying the pre-existing bug and fixing them both together later (assuming this is actually a bug and there's not some special case I haven't found documentation for)

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:15934]: Both instructions were added together. The currently defined feature is HasAtomicFlatPkAdd16Insts, so this mirrors that

**Review by rampitec — APPROVED** [review]
LGTM

**Comment by arsenm** [comment]
### Merge activity

* **Jun 15, 3:51 AM EDT**: @arsenm started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/95394).

**Comment by nico** [comment]
This or your other PR that landed at the same time probably caused http://45.33.8.238/linux/140692/step_11.txt

Please take a look and revert for now if it takes a while to fix.

**Comment by nico** [comment]
It was this one. Reverted in 95b77d90aae10725ea692e120aac083ef1c1297d for now.

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 3]
```diff
@@ -15942,6 +15942,16 @@ static bool isHalf2OrBFloat2(Type *Ty) {
   return false;
 }
 
+static bool isHalf2(Type *Ty) {
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:15945]: Can we make this `isV2F16` please? LLVM already has a couple of conventions for naming vector types, I don't think we need to add another one.

**Pull request merged**